cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519ML.hs

```diff
@@ -0,0 +1,278 @@
+-- are used via the 'MonadSodium' typeclass, which is responsible for
+-- guaranteeing orderly execution of these actions. We avoid using these
+-- primitives inside 'unsafeIOToST', as well as any 'IO' actions that would be
+-- unsafe to use inside 'unsafePerformIO'.
+--
+-- Specifically, we only use the following potentially dangerous operations
+-- inside 'unsafeIOToST':
+--
+-- - Libsodium functions marshalled from C, which do not allocate and only
+--   perform operations on previously allocated memory through stable pointers,
+--   and that are morally referentially transparent (that is, while they do
+--   perform destructive updates, the update is idempotent in the imperative
+--   sense, making it safe to execute repeatedly with the same input).
+-- - 'getErrno'; however, 'ST' guarantees sequentiality in the context where
+--   we use 'getErrno', so this is actually fine.
+-- - 'BS.useAsCStringLen', which is actually fine and shouldn't require 'IO'
+--   to begin with, but unfortunately it does.
+instance (MonadST m, MonadSodium m, MonadThrow m) => DSIGNMAlgorithm m Ed25519DSIGNM where
+    deriveVerKeyDSIGNM (SignKeyEd25519DSIGNM sk) =
+      VerKeyEd25519DSIGNM <$> do
+        mlsbUseAsSizedPtr sk $ \skPtr -> do
+          (psb, maybeErrno) <- withLiftST $ \fromST -> fromST $ do
+              psbCreateSizedResult $ \pkPtr ->
+                cOrError $ unsafeIOToST $
+                  c_crypto_sign_ed25519_sk_to_pk pkPtr skPtr
+          throwOnErrno "deriveVerKeyDSIGNM @Ed25519DSIGNM" "c_crypto_sign_ed25519_sk_to_pk" maybeErrno
+          return psb
+
+
+    signDSIGNM () a (SignKeyEd25519DSIGNM sk) =
+      let bs = getSignableRepresentation a
+      in SigEd25519DSIGNM <$> do
+          mlsbUseAsSizedPtr sk $ \skPtr -> do
+            (psb, maybeErrno) <- withLiftST $ \fromST -> fromST $ do
+                psbCreateSizedResult $ \sigPtr -> do
+                  cOrError $ unsafeIOToST $ do
+                    BS.useAsCStringLen bs $ \(ptr, len) ->
+                      c_crypto_sign_ed25519_detached sigPtr nullPtr (castPtr ptr) (fromIntegral len) skPtr
+            throwOnErrno "signDSIGNM @Ed25519DSIGNM" "c_crypto_sign_ed25519_detached" maybeErrno
+            return psb
+
+    --
+    -- Key generation
+    --
+    {-# NOINLINE genKeyDSIGNM #-}
+    genKeyDSIGNM seed = SignKeyEd25519DSIGNM <$> do
+          sk <- mlsbNew
+          mlsbUseAsSizedPtr sk $ \skPtr ->
+            mlockedSeedUseAsCPtr seed $ \seedPtr -> do
+              maybeErrno <- withLiftST $ \fromST ->
+                fromST $ allocaSizedST $ \pkPtr -> do
+                  cOrError $ unsafeIOToST $
+                    c_crypto_sign_ed25519_seed_keypair pkPtr skPtr (SizedPtr . castPtr $ seedPtr)
+              throwOnErrno "genKeyDSIGNM @Ed25519DSIGNM" "c_crypto_sign_ed25519_seed_keypair" maybeErrno
+          return sk
+
+    getSeedDSIGNM _ (SignKeyEd25519DSIGNM sk) = do
+        seed <- mlockedSeedNew
```

Comment:
The inconsistent coding styles are in fact bothering me a fair bit too; they are at least in part the result of many cooks handling this part of the codebase, and several changes being performed on it in parallel, and all the merging and rebasing and cherry-picking that goes with that. I'm not a huge fan of *enforcing* these things, but I am very much in favor of running one over the repo and committing that as a "whitespace" commit, even if it causes a lot of diff noise (worst case, `--ignore-all-space` should prevent the worst). Also happy to shave various yaks with you on the matter, if you so desire.

Reply:
Oh, but this particular case is probably the result of adding or removing wrapper functions, and I may have kept the wrapped block at the previous indentation level to minimize diff noise. But it is kind of ridiculous this way, they should both be 2 spaces like everything else in that module.

Reply:
No worries, I am not gonna pay attention too much to formatting in current PRs. I will be definitely enforcing formolu on the repo in not so distant future. This will remove the need to discuss it in the future period.